File: drafts/bs-Zarith/lib/js/src/Natural.js
'use strict';

var Bigint = require("./Bigint.js");
var Caml_exceptions = require("rescript/lib/js/caml_exceptions.js");
var Caml_js_exceptions = require("rescript/lib/js/caml_js_exceptions.js");

var Underflow = /* @__PURE__ */Caml_exceptions.create("Natural.Underflow");

var DecodeError = /* @__PURE__ */Caml_exceptions.create("Natural.DecodeError");

var zero = /* Natural */{
  _0: Bigint.zero
};

var one = /* Natural */{
  _0: Bigint.one
};

function ofInt(x) {
  var r = Bigint.of_int(x);
  if (Bigint.lt(r, Bigint.zero)) {
    return ;
  } else {
    return /* Natural */{
            _0: r
          };
  }
}

function ofInt32(x) {
  var r = Bigint.of_int32(x);
  if (Bigint.lt(r, Bigint.zero)) {
    return ;
  } else {
    return /* Natural */{
            _0: r
          };
  }
}

function ofInt64(x) {
  var r = Bigint.of_int64(x);
  if (Bigint.lt(r, Bigint.zero)) {
    return ;
  } else {
    return /* Natural */{
            _0: r
          };
  }
}

function ofFloat(x) {
  var r = Bigint.of_float(x);
  if (Bigint.lt(r, Bigint.zero)) {
    return ;
  } else {
    return /* Natural */{
            _0: r
          };
  }
}

function ofString(x) {
  var r = Bigint.of_string_opt(x);
  if (r !== undefined && !Bigint.lt(r, Bigint.zero)) {
    return /* Natural */{
            _0: r
          };
  }
  
}

function unsafeOfInt(x) {
  var x$1 = ofInt(x);
  if (x$1 !== undefined) {
    return x$1;
  }
  throw {
        RE_EXN_ID: Underflow,
        Error: new Error()
      };
}

function unsafeOfInt32(x) {
  var x$1 = ofInt32(x);
  if (x$1 !== undefined) {
    return x$1;
  }
  throw {
        RE_EXN_ID: Underflow,
        Error: new Error()
      };
}

function unsafeOfInt64(x) {
  var x$1 = ofInt64(x);
  if (x$1 !== undefined) {
    return x$1;
  }
  throw {
        RE_EXN_ID: Underflow,
        Error: new Error()
      };
}

function unsafeOfFloat(x) {
  var x$1 = ofFloat(x);
  if (x$1 !== undefined) {
    return x$1;
  }
  throw {
        RE_EXN_ID: Underflow,
        Error: new Error()
      };
}

function unsafeOfString(x) {
  var x$1 = ofString(x);
  if (x$1 !== undefined) {
    return x$1;
  }
  throw {
        RE_EXN_ID: DecodeError,
        Error: new Error()
      };
}

function toInt(x) {
  return Bigint.to_int(x._0);
}

function toInt32(x) {
  return Bigint.to_int32(x._0);
}

function toInt64(x) {
  return Bigint.to_int64(x._0);
}

function toFloat(x) {
  return Bigint.to_float(x._0);
}

function toString(x) {
  return Bigint.to_string(x._0);
}

function succ(x) {
  return /* Natural */{
          _0: Bigint.add(x._0, Bigint.one)
        };
}

function pred(x) {
  var r = Bigint.sub(x._0, Bigint.one);
  if (Bigint.lt(r, Bigint.zero)) {
    throw {
          RE_EXN_ID: Underflow,
          Error: new Error()
        };
  }
  return /* Natural */{
          _0: r
        };
}

function predSafe(x) {
  var x$1;
  try {
    x$1 = pred(x);
  }
  catch (raw_exn){
    var exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.RE_EXN_ID === Underflow) {
      return ;
    }
    throw exn;
  }
  return x$1;
}

function add(x, y) {
  return /* Natural */{
          _0: Bigint.add(x._0, y._0)
        };
}

function sub(x, y) {
  var r = Bigint.sub(x._0, y._0);
  if (Bigint.lt(r, Bigint.zero)) {
    throw {
          RE_EXN_ID: Underflow,
          Error: new Error()
        };
  }
  return /* Natural */{
          _0: r
        };
}

function subSafe(x, y) {
  var r;
  try {
    r = sub(x, y);
  }
  catch (raw_exn){
    var exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    if (exn.RE_EXN_ID === Underflow) {
      return ;
    }
    throw exn;
  }
  return r;
}

function mul(x, y) {
  return /* Natural */{
          _0: Bigint.mul(x._0, y._0)
        };
}

function div(x, y) {
  return /* Natural */{
          _0: Bigint.div(x._0, y._0)
        };
}

function rem(x, y) {
  return /* Natural */{
          _0: Bigint.rem(x._0, y._0)
        };
}

function divRem(x, y) {
  var match = Bigint.div_rem(x._0, y._0);
  return [
          /* Natural */{
            _0: match[0]
          },
          /* Natural */{
            _0: match[1]
          }
        ];
}

function cdiv(x, y) {
  return /* Natural */{
          _0: Bigint.cdiv(x._0, y._0)
        };
}

function fdiv(x, y) {
  return /* Natural */{
          _0: Bigint.fdiv(x._0, y._0)
        };
}

function edivRem(x, y) {
  var match = Bigint.ediv_rem(x._0, y._0);
  return [
          /* Natural */{
            _0: match[0]
          },
          /* Natural */{
            _0: match[1]
          }
        ];
}

function ediv(x, y) {
  return /* Natural */{
          _0: Bigint.ediv(x._0, y._0)
        };
}

function erem(x, y) {
  return /* Natural */{
          _0: Bigint.erem(x._0, y._0)
        };
}

function divexact(x, y) {
  return /* Natural */{
          _0: Bigint.divexact(x._0, y._0)
        };
}

exports.Underflow = Underflow;
exports.DecodeError = DecodeError;
exports.zero = zero;
exports.one = one;
exports.ofInt = ofInt;
exports.ofInt32 = ofInt32;
exports.ofInt64 = ofInt64;
exports.ofFloat = ofFloat;
exports.ofString = ofString;
exports.unsafeOfInt = unsafeOfInt;
exports.unsafeOfInt32 = unsafeOfInt32;
exports.unsafeOfInt64 = unsafeOfInt64;
exports.unsafeOfFloat = unsafeOfFloat;
exports.unsafeOfString = unsafeOfString;
exports.toInt = toInt;
exports.toInt32 = toInt32;
exports.toInt64 = toInt64;
exports.toFloat = toFloat;
exports.toString = toString;
exports.succ = succ;
exports.pred = pred;
exports.predSafe = predSafe;
exports.add = add;
exports.sub = sub;
exports.subSafe = subSafe;
exports.mul = mul;
exports.div = div;
exports.rem = rem;
exports.divRem = divRem;
exports.cdiv = cdiv;
exports.fdiv = fdiv;
exports.edivRem = edivRem;
exports.ediv = ediv;
exports.erem = erem;
exports.divexact = divexact;
/* No side effect */

File: drafts/bs-Zarith/src/Natural.ml
(* 
 * Module:     Natural
 * Copyright:  Plow Technologies, 2020
 * Maintainer: james.haver@plowtech.net 
 *
 * Support arbitrary precision Natural numbers in OCaml.
 *)

exception Underflow
exception DecodeError

type t = Natural of Bigint.t

let zero = Natural Bigint.zero
let one = Natural Bigint.one

(* of conversion *)

let ofInt x =
  let r = Bigint.of_int x in
  if Bigint.lt r Bigint.zero
  then None
  else Some (Natural r)

let ofInt32 x =
  let r = Bigint.of_int32 x in
  if Bigint.lt r Bigint.zero
  then None
  else Some (Natural r)

let ofInt64 x =
  let r = Bigint.of_int64 x in
  if Bigint.lt r Bigint.zero
  then None
  else Some (Natural r)

let ofFloat x =
  let r = Bigint.of_float x in
  if Bigint.lt r Bigint.zero
  then None
  else Some (Natural r)
  
let ofString x =
  let r = Bigint.of_string_opt x in
  match r with
  | None -> None
  | Some r ->
     if Bigint.lt r Bigint.zero
     then None
     else Some (Natural r)

let unsafeOfInt x =
  match ofInt x with
  | Some x -> x
  | None -> raise Underflow

let unsafeOfInt32 x =
  match ofInt32 x with
  | Some x -> x
  | None -> raise Underflow

let unsafeOfInt64 x =
  match ofInt64 x with
  | Some x -> x
  | None -> raise Underflow

let unsafeOfFloat x =
  match ofFloat x with
  | Some x -> x
  | None -> raise Underflow

let unsafeOfString x =
  match ofString x with
  | Some x -> x
  | None -> raise DecodeError
          
(* to conversion *)

let toInt x =
  match x with
  | Natural x -> Bigint.to_int x

let toInt32 x =
  match x with
  | Natural x -> Bigint.to_int32 x

let toInt64 x =
  match x with
  | Natural x -> Bigint.to_int64 x

let toFloat x =
  match x with
  | Natural x -> Bigint.to_float x

let toString x =
  match x with
  | Natural x -> Bigint.to_string x
               
(* Arithmetic *)        
let succ x =
  match x with
  | Natural x -> Natural (Bigint.add x Bigint.one)

let pred x =
  match x with
  | Natural x ->
     let r = Bigint.sub x Bigint.one in
     if Bigint.lt r Bigint.zero
     then raise Underflow
     else Natural r

let predSafe x =
  match pred x with
  | x -> Some x
  | exception Underflow -> None
     
let add x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.add x y)

let sub x y =
  match (x,y) with
  | (Natural x, Natural y) ->
     let r = Bigint.sub x y in
     if Bigint.lt r Bigint.zero
     then raise Underflow
     else Natural r

let subSafe x y =
  match sub x y with
  | r -> Some r
  | exception Underflow -> None
     
let mul x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.mul x y)

let div x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.div x y)

let rem x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.rem x y)

let divRem x y =
  match (x,y) with
  | (Natural x, Natural y) ->
     let (r1,r2) = (Bigint.div_rem x y) in
     (Natural r1, Natural r2)
                            
let cdiv x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.cdiv x y)

let fdiv x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.fdiv x y)
                            
let edivRem x y =
  match (x,y) with
  | (Natural x, Natural y) ->
     let (r1,r2) = (Bigint.ediv_rem x y) in
     (Natural r1, Natural r2)

let ediv x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.ediv x y)

let erem x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.erem x y)

let divexact x y =
  match (x,y) with
  | (Natural x, Natural y) -> Natural (Bigint.divexact x y)
